### ISSUE TITLE
DTLS-SCTP Association lifetime

### ISSUE THREAD

**Issue by gloinul** [gloinul/draft-westerlund-tsvwg-dtls-over-sctp-bis#63]
We may need text for the fact that DTLS 1.3 is limited to 65535 rekeyings and then the DTLS assocation needs to be terminated and restarted in that case. This produces a limited association lifetime that may need to be made explicit in the document.

**Comment by tuexen** [comment]
Maybe add to section 1.1.2.

Using DTLS 1.2 or DTLS 1.3 instead to using DTLS 1.0 limits the lifetime of a DTLS connection and the data volume which can be transferred over a DTLS  connection. Especially:
* The number of renegotations in DTLS 1.2 is limited to 65534 compared to unlimited in DTLS 1.0.
* The number of KeyUpdates in DTLS 1.3 is limited to 65532 and renegotiations are not supported.

**Comment by gloinul** [comment]
Yes, I think that would make sense to add. Can you (@tuexen) create a PR for it?

**Comment by tuexen** [comment]
Yepp, give me an hour or so...

**Comment by emanjon** [comment]
Following the ANSSI recommendation and doing one rekeying once per hour it takes 7.5 years to reach epcoh 2^16 so that will not be a problem. But rekeying after 2^24.5 records could maybe be a problem. You can only send 2^40.5 DTLS AES-GCM records on a connection in DTLS 1.3.

**Comment by gloinul** [comment]
So the maximum amount of data that can be sent is then on the order of  2^56.5 bytes under that ciphers limitation as we can at least send 2^16-1 bytes per DTLS record, happy we are not using IP MTU DTLS records.

**Comment by gloinul** [comment]
So #64 has been merged, but lets think if we need in the next version to relate those numbers to actual limitation discussed in this thread?

**Comment by tuexen** [comment]
> So the maximum amount of data that can be sent is then on the order of 2^56.5 bytes under that ciphers limitation as we can at least send 2^16-1 bytes per DTLS record, happy we are not using IP MTU DTLS records.

You are correct, but in case the application is sending small user messages you end up using small DTLS records...

**Comment by tuexen** [comment]
> Following the ANSSI recommendation and doing one rekeying once per hour it takes 7.5 years to reach epcoh 2^16 so that will not be a problem. But rekeying after 2^24.5 records could maybe be a problem. You can only send 2^40.5 DTLS AES-GCM records on a connection in DTLS 1.3.

Don't the same limits apply for DTLS 1.2 when using the same crypto algorithms? Just using renegotiations instead of key updates?

**Comment by tuexen** [comment]
> So #64 has been merged, but lets think if we need in the next version to relate those numbers to actual limitation discussed in this thread?

Yepp, makes sense...

**Comment by emanjon** [comment]
I made a DTLS 1.3 issue of this. Seem quite likely that it will be fixed

https://github.com/tlswg/dtls13-spec/issues/249

**Comment by emanjon** [comment]
I made a PR for DTLS 1.3

https://github.com/tlswg/dtls13-spec/pull/250

@gloinul @tuexen Is uint32 epoch enough? I.e. 2^56.5 AES-GCM records?

**Comment by tuexen** [comment]
Is there are reason why the epoch can't wrap around?

**Comment by gloinul** [comment]
So uint32 epoch would be sufficient for time and total number of bytes when using large messages. So lets check if there is an issue if that each user message consumes one DTLS record. 

So if an application sends a 200 byte user message (assume 250 bytes with overhead) and sends a lot of them to fill a 1 GBPS link. Then we can send 500,000 messages per second which is almost 2^19 messages. Which would deplete the range in 2^37.5, which is 6000 years. So if the user messages are 1 byte + overhead (25 bytes?) and the processing is capable of keeping up with 10 Gbps then we are down in the range of  below 10 years. So maybe to be certain a uint64 would be safer. As it is not needed to be sent, and if there are no limitations that is relevant for number of key derivations then I don't see an issue with going safe and declare it a unit64 instead.

**Comment by gloinul** [comment]
So if we are going with outlined solution in #59 for how to deal with epoch changes, then wrap around is not the best, and we would prefer a sufficient long epoch. unit64 would be better to ensure long term safety, even if uint32 likely is sufficient. 

When the DTLS spec has been updated we need to reflect these changes in our spec so that limitations are correctly described.

**Comment by emanjon** [comment]
DTLS 1.3 will go for a modulo solution with a 64 bit epoch of which 16 bits used in the record plaintext

https://github.com/tlswg/dtls13-spec/pull/255

This PR has been approved but is not yet merged.
